modin/engines/base/frame/data.py

```diff
@@ -1719,79 +1765,71 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
-                    axis, self._partitions, lambda df: df
-                )
-                other[i]._partitions = other[i]._frame_mgr_cls.map_axis_partitions(
-                    axis, other[i]._partitions, lambda df: df
-                )
+        # define helper functions
+        def get_column_widths(partitions):
+            if len(partitions) > 0:
+                return [obj.width() for obj in partitions[0]]
 
-        if (
-            all(o.axes[axis].equals(self.axes[axis]) for o in other)
-            and not is_aligning_applied
-        ):
-            return (
-                self._partitions,
-                [self._simple_shuffle(axis, o) for o in other],
-                self.axes[axis].copy(),
-            )
+        def get_row_lengths(partitions):
+            if len(partitions.T) > 0:
+                return [obj.length() for obj in partitions.T[0]]
 
-        index_other_obj = [o.axes[axis] for o in other]
-        joined_index = self._join_index_objects(axis, index_other_obj, how, sort)
-        # We have to set these because otherwise when we perform the functions it may
-        # end up serializing this entire object.
-        left_old_idx = self.axes[axis]
-        right_old_idxes = index_other_obj
+        self_index = self.axes[axis]
+        others_index = [o.axes[axis] for o in other]
+        joined_index, make_reindexer = self._join_index_objects(
+            axis, [self_index] + others_index, how, sort
+        )
 
-        def make_map_func():
-            if not joined_index.is_unique and axis == 0:
-                return lambda df: df
-            return lambda df: df.reindex(joined_index, axis=axis)
+        # define conditions for reindexing and repartitioning `self` frame
+        do_reindex_self = not self_index.equals(joined_index)
+        do_repartition_self = force_repartition or do_reindex_self
 
-        # Start with this and we'll repartition the first time, and then not again.
-        if is_aligning_applied or (
-            not force_repartition and left_old_idx.equals(joined_index)
-        ):
-            reindexed_self = self._partitions
-        else:
+        # perform repartitioning and reindexing for `self` frame if needed
+        if do_repartition_self:
             reindexed_self = self._frame_mgr_cls.map_axis_partitions(
                 axis,
                 self._partitions,
-                make_map_func(),
+                # self frame has 0 idx
+                make_reindexer(do_reindex_self, 0),
             )
+        else:
+            reindexed_self = self._partitions
 
-        def get_column_widths(partitions):
-            if len(partitions) > 0:
-                return [obj.width() for obj in partitions[0]]
+        # define length of `self` and `other` frames to aligning purpose
+        self_lengths = (
+            get_row_lengths(reindexed_self)
+            if axis == 0
+            else get_column_widths(reindexed_self)
```

Comment:
I think we can encapsulate this logic inside the appropriate function by adding `axis` parameter:
```python
get_axis_widths(reindexed_self, axis)
```